File: software/mobile/android/core-update/src/main/kotlin/com/avlyubimov/svc/core/update/FirmwareReleaseRepository.kt
package com.avlyubimov.svc.core.update

import com.avlyubimov.svc.core.protocol.FirmwareComponent
import com.avlyubimov.svc.core.protocol.FirmwareManifest
import com.avlyubimov.svc.core.protocol.FirmwareManifestParser
import com.avlyubimov.svc.core.protocol.SignatureVerifier
import com.avlyubimov.svc.core.protocol.validateArtifact
import java.net.HttpURLConnection
import java.net.URI
import java.security.KeyFactory
import java.security.MessageDigest
import java.security.Signature
import java.security.spec.MGF1ParameterSpec
import java.security.spec.PSSParameterSpec
import java.security.spec.X509EncodedKeySpec
import java.util.Base64
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

enum class FirmwareReleaseChannel(val manifestChannel: String) {
    STABLE("stable"),
    BETA("beta"),
    TEST("dev"),
}

interface FirmwareReleaseRepository {
    suspend fun fetchManifest(
        channel: FirmwareReleaseChannel = FirmwareReleaseChannel.STABLE,
    ): FirmwareManifest
}

class MockFirmwareReleaseRepository : FirmwareReleaseRepository {
    override suspend fun fetchManifest(channel: FirmwareReleaseChannel): FirmwareManifest =
        FirmwareManifest(
            schemaVersion = 1,
            releaseVersion = "0.1.0-test.1",
            releaseTag = "svc-v0.1.0-test.1",
            channel = "dev",
            minimumMobileVersion = "0.1.0",
            components = emptyList(),
        )
}

data class ReleaseHttpResponse(
    val statusCode: Int,
    val body: ByteArray,
)

fun interface ReleaseHttpClient {
    fun get(uri: URI, accept: String): ReleaseHttpResponse
}

class UrlConnectionReleaseHttpClient : ReleaseHttpClient {
    override fun get(uri: URI, accept: String): ReleaseHttpResponse {
        require(uri.scheme == "https") { "firmware release API requires HTTPS" }
        val connection = uri.toURL().openConnection() as HttpURLConnection
        connection.connectTimeout = 10_000
        connection.readTimeout = 30_000
        connection.instanceFollowRedirects = true
        connection.setRequestProperty("Accept", accept)
        connection.setRequestProperty("X-GitHub-Api-Version", "2022-11-28")
        val status = connection.responseCode
        val stream = if (status in 200..299) connection.inputStream else connection.errorStream
        return ReleaseHttpResponse(status, stream?.use { it.readBytes() } ?: byteArrayOf())
    }
}

class RsaPssReleaseSignatureVerifier(
    subjectPublicKeyInfoPem: String,
    private val allowedKeyId: String,
) : SignatureVerifier {
    private val publicKey = KeyFactory.getInstance("RSA").generatePublic(
        X509EncodedKeySpec(
            Base64.getDecoder().decode(
                subjectPublicKeyInfoPem
                    .replace("-----BEGIN PUBLIC KEY-----", "")
                    .replace("-----END PUBLIC KEY-----", "")
                    .filterNot(Char::isWhitespace),
            ),
        ),
    )

    override fun verify(keyId: String, bytes: ByteArray, signature: ByteArray): Boolean {
        if (keyId != allowedKeyId) {
            return false
        }
        val verifier = Signature.getInstance("RSASSA-PSS")
        verifier.initVerify(publicKey)
        verifier.setParameter(
            PSSParameterSpec(
                "SHA-256",
                "MGF1",
                MGF1ParameterSpec.SHA256,
                32,
                1,
            ),
        )
        verifier.update(bytes)
        return verifier.verify(signature)
    }
}

class GitHubFirmwareReleaseRepository(
    private val owner: String,
    private val repository: String,
    private val signatureVerifier: SignatureVerifier,
    private val httpClient: ReleaseHttpClient = UrlConnectionReleaseHttpClient(),
    private val parser: FirmwareManifestParser = FirmwareManifestParser(),
    private val json: Json = Json { ignoreUnknownKeys = true },
    private val apiBaseUri: URI = URI("https://api.github.com"),
) : FirmwareReleaseRepository {
    override suspend fun fetchManifest(
        channel: FirmwareReleaseChannel,
    ): FirmwareManifest = withContext(Dispatchers.IO) {
        val release = fetchRelease(channel)
        val manifestAsset = release.asset("firmware-manifest.json")
        val manifestBytes = download(manifestAsset)
        val manifest = parser.parse(manifestBytes.decodeToString())
        validateReleaseIdentity(manifest, release.tagName, channel)
        manifest.components.forEach { component ->
            val image = download(release.asset(component.file))
            val signature = download(release.asset(component.releaseSignature.file))
            validateArtifact(component, image, signature, signatureVerifier)
        }
        manifest
    }

    private fun fetchRelease(channel: FirmwareReleaseChannel): GitHubRelease {
        if (channel == FirmwareReleaseChannel.STABLE) {
            val uri = apiBaseUri.resolve("/repos/$owner/$repository/releases/latest")
            return json.decodeFromString<GitHubRelease>(
                requestJson(uri).decodeToString(),
            ).also {
                require(!it.draft && !it.prerelease) {
                    "stable release metadata is invalid"
                }
            }
        }
        val uri = apiBaseUri.resolve("/repos/$owner/$repository/releases?per_page=100")
        return json.decodeFromString<List<GitHubRelease>>(requestJson(uri).decodeToString())
            .firstOrNull {
                !it.draft &&
                    it.prerelease &&
                    releaseTagMatches(it.tagName, channel)
            }
            ?: throw FirmwareReleaseException("release unavailable")
    }

    private fun requestJson(uri: URI): ByteArray {
        val response = httpClient.get(uri, "application/vnd.github+json")
        require(response.statusCode in 200..299) { "GitHub API request failed" }
        return response.body
    }

    private fun download(asset: GitHubReleaseAsset): ByteArray {
        val uri = URI(asset.browserDownloadUrl)
        require(uri.scheme == "https" && uri.host == "github.com") {
            "untrusted release asset URL"
        }
        val response = httpClient.get(uri, "application/octet-stream")
        require(response.statusCode in 200..299) { "asset download failed" }
        require(response.body.size.toLong() == asset.size) { "asset size mismatch" }
        require(asset.digest == "sha256:${sha256(response.body)}") {
            "asset SHA-256 mismatch"
        }
        return response.body
    }

    private fun validateReleaseIdentity(
        manifest: FirmwareManifest,
        releaseTag: String,
        requestedChannel: FirmwareReleaseChannel,
    ) {
        require(manifest.releaseTag == releaseTag) { "release tag mismatch" }
        require(manifest.releaseTag == "svc-v${manifest.releaseVersion}") {
            "release tag/version mismatch"
        }
        require(manifest.channel == requestedChannel.manifestChannel) {
            "release channel mismatch"
        }
        require(releaseTagMatches(releaseTag, requestedChannel)) {
            "release tag/channel mismatch"
        }
        require(manifest.components.all { it.version == manifest.releaseVersion }) {
            "component version mismatch"
        }
    }

    private fun releaseTagMatches(tag: String, channel: FirmwareReleaseChannel): Boolean {
        val core = "(?:0|[1-9][0-9]*)\\.(?:0|[1-9][0-9]*)\\.(?:0|[1-9][0-9]*)"
        val pattern = when (channel) {
            FirmwareReleaseChannel.STABLE -> "^svc-v$core$"
            FirmwareReleaseChannel.BETA -> "^svc-v$core-beta\\.(?:0|[1-9][0-9]*)$"
            FirmwareReleaseChannel.TEST -> "^svc-v$core-test\\.(?:0|[1-9][0-9]*)$"
        }
        return Regex(pattern).matches(tag)
    }

    private fun sha256(bytes: ByteArray): String =
        MessageDigest.getInstance("SHA-256")
            .digest(bytes)
            .joinToString("") { "%02x".format(it) }
}

@Serializable
private data class GitHubRelease(
    @SerialName("tag_name")
    val tagName: String,
    val draft: Boolean,
    val prerelease: Boolean,
    val assets: List<GitHubReleaseAsset>,
) {
    fun asset(name: String): GitHubReleaseAsset =
        assets.singleOrNull { it.name == name }
            ?: throw FirmwareReleaseException("release asset missing: $name")
}

@Serializable
private data class GitHubReleaseAsset(
    val name: String,
    val size: Long,
    @SerialName("browser_download_url")
    val browserDownloadUrl: String,
    val digest: String,
)

class FirmwareReleaseException(message: String) : IllegalArgumentException(message)

File: software/mobile/ios/SVCMobile/Sources/Data/FirmwareReleaseRepository.swift
import CryptoKit
import Foundation

enum FirmwareReleaseChannel: String {
    case stable
    case beta
    case test

    var manifestChannel: String {
        self == .test ? "dev" : rawValue
    }
}

protocol FirmwareReleaseRepository {
    func fetchManifest(channel: FirmwareReleaseChannel) async throws -> FirmwareManifest
}

struct MockFirmwareReleaseRepository: FirmwareReleaseRepository {
    func fetchManifest(channel: FirmwareReleaseChannel) async throws -> FirmwareManifest {
        FirmwareManifest(
            schemaVersion: 1,
            releaseVersion: "0.1.0-test.1",
            releaseTag: "svc-v0.1.0-test.1",
            channel: "dev",
            minimumMobileVersion: "0.1.0",
            components: []
        )
    }
}

protocol FirmwareReleaseSignatureVerifying {
    func verify(
        data: Data,
        signature: Data,
        keyId: String,
        algorithm: String
    ) -> Bool
}

struct GitHubFirmwareReleaseRepository: FirmwareReleaseRepository {
    let owner: String
    let repository: String
    let signatureVerifier: FirmwareReleaseSignatureVerifying
    var session: URLSession = .shared
    var apiBaseURL = URL(string: "https://api.github.com")!

    func fetchManifest(channel: FirmwareReleaseChannel) async throws -> FirmwareManifest {
        let release = try await fetchRelease(channel: channel)
        let manifestAsset = try asset(named: "firmware-manifest.json", in: release)
        let manifestData = try await download(manifestAsset)
        let manifest = try JSONDecoder().decode(FirmwareManifest.self, from: manifestData)
        try validate(
            manifest: manifest,
            releaseTag: release.tagName,
            requestedChannel: channel
        )

        for component in manifest.components {
            let imageAsset = try asset(named: component.file, in: release)
            let signatureAsset = try asset(
                named: component.releaseSignature.file,
                in: release
            )
            let image = try await download(imageAsset)
            let signature = try await download(signatureAsset)
            try validate(
                component: component,
                image: image,
                signature: signature
            )
        }
        return manifest
    }

    private func fetchRelease(channel: FirmwareReleaseChannel) async throws -> GitHubRelease {
        switch channel {
        case .stable:
            let url = apiBaseURL
                .appending(path: "repos")
                .appending(path: owner)
                .appending(path: repository)
                .appending(path: "releases")
                .appending(path: "latest")
            let release = try await request(url: url, as: GitHubRelease.self)
            guard !release.draft, !release.prerelease else {
                throw FirmwareReleaseError.releaseIdentityMismatch
            }
            return release
        case .beta, .test:
            var components = URLComponents(
                url: apiBaseURL
                    .appending(path: "repos")
                    .appending(path: owner)
                    .appending(path: repository)
                    .appending(path: "releases"),
                resolvingAgainstBaseURL: false
            )
            components?.queryItems = [URLQueryItem(name: "per_page", value: "100")]
            guard let url = components?.url else {
                throw FirmwareReleaseError.invalidResponse
            }
            let releases = try await request(url: url, as: [GitHubRelease].self)
            guard let release = releases.first(where: {
                !$0.draft &&
                    $0.prerelease &&
                    releaseTag($0.tagName, matches: channel)
            }) else {
                throw FirmwareReleaseError.releaseUnavailable
            }
            return release
        }
    }

    private func request<Value: Decodable>(
        url: URL,
        as type: Value.Type
    ) async throws -> Value {
        var request = URLRequest(url: url)
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        request.setValue("2022-11-28", forHTTPHeaderField: "X-GitHub-Api-Version")
        let (data, response) = try await session.data(for: request)
        try requireSuccess(response)
        return try JSONDecoder().decode(type, from: data)
    }

    private func download(_ asset: GitHubReleaseAsset) async throws -> Data {
        guard
            asset.browserDownloadURL.scheme == "https",
            asset.browserDownloadURL.host == "github.com"
        else {
            throw FirmwareReleaseError.untrustedAssetURL
        }
        var request = URLRequest(url: asset.browserDownloadURL)
        request.setValue("application/octet-stream", forHTTPHeaderField: "Accept")
        let (data, response) = try await session.data(for: request)
        try requireSuccess(response)
        guard data.count == asset.size else {
            throw FirmwareReleaseError.sizeMismatch
        }
        guard asset.digest == "sha256:\(sha256(data))" else {
            throw FirmwareReleaseError.sha256Mismatch
        }
        return data
    }

    private func validate(
        manifest: FirmwareManifest,
        releaseTag: String,
        requestedChannel: FirmwareReleaseChannel
    ) throws {
        guard manifest.schemaVersion == 1 else {
            throw FirmwareReleaseError.unsupportedManifest
        }
        guard
            manifest.releaseTag == releaseTag,
            manifest.releaseTag == "svc-v\(manifest.releaseVersion)",
            manifest.channel == requestedChannel.manifestChannel,
            self.releaseTag(releaseTag, matches: requestedChannel),
            !manifest.components.isEmpty,
            Set(manifest.components.map(\.target)).count == manifest.components.count,
            manifest.components.allSatisfy({ $0.version == manifest.releaseVersion })
        else {
            throw FirmwareReleaseError.releaseIdentityMismatch
        }
        for component in manifest.components {
            guard
                ["stm32-main", "e73-radio"].contains(component.target),
                component.size >= 0,
                component.releaseSignature.size > 0,
                isAssetBasename(component.file),
                isAssetBasename(component.releaseSignature.file),
                component.file != component.releaseSignature.file,
                component.releaseSignature.algorithm == "rsa-pss-sha256"
            else {
                throw FirmwareReleaseError.unsupportedManifest
            }
            let expectedFormat = component.target == "e73-radio"
                ? "mcuboot-signed"
                : "oemirot-signed"
            guard
                !(component.imageFormat == "review-raw" && component.installable),
                !component.installable || component.imageFormat == expectedFormat
            else {
                throw FirmwareReleaseError.unsupportedManifest
            }
        }
    }

    private func validate(
        component: FirmwareComponent,
        image: Data,
        signature: Data
    ) throws {
        guard image.count == component.size else {
            throw FirmwareReleaseError.sizeMismatch
        }
        guard sha256(image) == component.sha256 else {
            throw FirmwareReleaseError.sha256Mismatch
        }
        guard signature.count == component.releaseSignature.size else {
            throw FirmwareReleaseError.signatureSizeMismatch
        }
        guard sha256(signature) == component.releaseSignature.sha256 else {
            throw FirmwareReleaseError.signatureSha256Mismatch
        }
        guard signatureVerifier.verify(
            data: image,
            signature: signature,
            keyId: component.releaseSignature.keyId,
            algorithm: component.releaseSignature.algorithm
        ) else {
            throw FirmwareReleaseError.invalidSignature
        }
    }

    private func asset(
        named name: String,
        in release: GitHubRelease
    ) throws -> GitHubReleaseAsset {
        guard let asset = release.assets.first(where: { $0.name == name }) else {
            throw FirmwareReleaseError.assetMissing
        }
        return asset
    }

    private func requireSuccess(_ response: URLResponse) throws {
        guard
            let response = response as? HTTPURLResponse,
            (200..<300).contains(response.statusCode)
        else {
            throw FirmwareReleaseError.invalidResponse
        }
    }

    private func releaseTag(
        _ tag: String,
        matches channel: FirmwareReleaseChannel
    ) -> Bool {
        let core = #"(?:0|[1-9][0-9]*)\.(?:0|[1-9][0-9]*)\.(?:0|[1-9][0-9]*)"#
        let pattern: String
        switch channel {
        case .stable:
            pattern = #"^svc-v\#(core)$"#
        case .beta:
            pattern = #"^svc-v\#(core)-beta\.(?:0|[1-9][0-9]*)$"#
        case .test:
            pattern = #"^svc-v\#(core)-test\.(?:0|[1-9][0-9]*)$"#
        }
        return tag.range(of: pattern, options: .regularExpression) != nil
    }

    private func isAssetBasename(_ value: String) -> Bool {
        value.range(
            of: #"^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$"#,
            options: .regularExpression
        ) != nil
    }

    private func sha256(_ data: Data) -> String {
        SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }
}

private struct GitHubRelease: Decodable {
    let tagName: String
    let draft: Bool
    let prerelease: Bool
    let assets: [GitHubReleaseAsset]

    enum CodingKeys: String, CodingKey {
        case tagName = "tag_name"
        case draft
        case prerelease
        case assets
    }
}

private struct GitHubReleaseAsset: Decodable {
    let name: String
    let size: Int
    let browserDownloadURL: URL
    let digest: String

    enum CodingKeys: String, CodingKey {
        case name
        case size
        case browserDownloadURL = "browser_download_url"
        case digest
    }
}

enum FirmwareReleaseError: Error, Equatable {
    case httpsRequired
    case invalidResponse
    case releaseUnavailable
    case assetMissing
    case untrustedAssetURL
    case unsupportedManifest
    case releaseIdentityMismatch
    case sizeMismatch
    case sha256Mismatch
    case signatureSizeMismatch
    case signatureSha256Mismatch
    case invalidSignature
}
